In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('Expert_Opinions.csv')

In [3]:
df = df.drop([0],axis=0) # deleting the heading row

In [4]:
df.head()

Unnamed: 0        Unnamed: 1    0 Unnamed: 3 Unnamed: 4  Unnamed: 5  \
1         NaN  6/2/2018 0:17:49  NaN   Option 2       dsds         1.0   
2         NaN  6/2/2018 0:21:27  NaN   Option 1        edc         2.0   
3         NaN  6/2/2018 0:17:49  NaN   Option 2       dsds         1.0   
4         NaN  6/2/2018 0:21:27  NaN   Option 1        edc         2.0   
5         NaN  6/2/2018 0:21:27  NaN   Option 1        edc         1.0   

   Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9      ...       Unnamed: 122  \
1         1.0         2.0         2.0         2.0      ...                1.0   
2         2.0         3.0         2.0         2.0      ...                2.0   
3         1.0         2.0         2.0         3.0      ...                1.0   
4         2.0         3.0         2.0         2.0      ...                2.0   
5         2.0         3.0         3.0         3.0      ...                2.0   

   Unnamed: 123  Unnamed: 124  Unnamed: 125  Unnamed: 126 Unnamed: 127  \
1           2.0           2.0           3.0           2.0          3.0   
2           3.0           1.0           1.0           2.0          1.0   
3           2.0           2.0           3.0           2.0          3.0   
4           3.0           1.0           1.0           2.0          1.0   
5           3.0           1.0           1.0           2.0          1.0   

  Unnamed: 128 Unnamed: 129  Unnamed: 130  Unnamed: 131  
1          3.0          3.0           1.0           1.0  
2          2.0          1.0           1.0           2.0  
3          3.0          3.0           1.0           1.0  
4          2.0          1.0           1.0           2.0  
5          2.0          1.0           1.0           2.0  

[5 rows x 132 columns]

Each row represents an **Expert** entry and each column represents a **Participant** entry.

In [5]:
df.columns

Index(['Unnamed: 0', 'Unnamed: 1', '0', 'Unnamed: 3', 'Unnamed: 4',
       'Unnamed: 5', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9',
       ...
       'Unnamed: 122', 'Unnamed: 123', 'Unnamed: 124', 'Unnamed: 125',
       'Unnamed: 126', 'Unnamed: 127', 'Unnamed: 128', 'Unnamed: 129',
       'Unnamed: 130', 'Unnamed: 131'],
      dtype='object', length=132)

In [6]:
df = df.drop(['Unnamed: 1'],axis=1) # deleting the timestamp column

In [7]:
df = df.dropna(axis=1)

In [8]:
df.head()

Unnamed: 3 Unnamed: 4  Unnamed: 5  Unnamed: 6  Unnamed: 7  Unnamed: 8  \
1   Option 2       dsds         1.0         1.0         2.0         2.0   
2   Option 1        edc         2.0         2.0         3.0         2.0   
3   Option 2       dsds         1.0         1.0         2.0         2.0   
4   Option 1        edc         2.0         2.0         3.0         2.0   
5   Option 1        edc         1.0         2.0         3.0         3.0   

   Unnamed: 9  Unnamed: 10  Unnamed: 11  Unnamed: 12      ...       \
1         2.0          1.0          1.0          1.0      ...        
2         2.0          2.0          2.0          3.0      ...        
3         3.0          1.0          1.0          1.0      ...        
4         2.0          2.0          2.0          3.0      ...        
5         3.0          2.0          2.0          3.0      ...        

   Unnamed: 122  Unnamed: 123 Unnamed: 124 Unnamed: 125  Unnamed: 126  \
1           1.0           2.0          2.0          3.0           2.0   
2           2.0           3.0          1.0          1.0           2.0   
3           1.0           2.0          2.0          3.0           2.0   
4           2.0           3.0          1.0          1.0           2.0   
5           2.0           3.0          1.0          1.0           2.0   

   Unnamed: 127  Unnamed: 128  Unnamed: 129  Unnamed: 130  Unnamed: 131  
1           3.0           3.0           3.0           1.0           1.0  
2           1.0           2.0           1.0           1.0           2.0  
3           3.0           3.0           3.0           1.0           1.0  
4           1.0           2.0           1.0           1.0           2.0  
5           1.0           2.0           1.0           1.0           2.0  

[5 rows x 120 columns]

So, we see that there are 120 columns. This is because there are 10 participants and 12 responses for each participants. The breakdown of questions are:
- 1 question for choosing the better mockup.
- 1 question for filling in the comments.
- 5 questions with 2 subparts (one for each option).

#### Creating meaningful columns

In [9]:
# Setting the constants
numQues = 5
colsPerParticipant = 2*numQues + 2
numParticipants = int(len(df.columns)/colsPerParticipant)

In [10]:
ques = ['Q' + str(i) for i in range(1,numQues+1)]
part = ['P'+str(i) for i in range(1,numParticipants+1)]
opt = ['O1','O2']
design = [part[i]+'D' for i in range(len(part))]
comments = [part[i]+'C' for i in range(len(part))] 

In [11]:
comments

['P1C', 'P2C', 'P3C', 'P4C', 'P5C', 'P6C', 'P7C', 'P8C', 'P9C', 'P10C']

In [12]:
tmp = [part[j] + ques[i] for j in range(len(part)) for i in range(len(ques))]
tmp = [tmp[i] + opt[j] for i in range(len(tmp)) for j in range(len(opt))]

In [13]:
final = []
k = 0
for i in range(len(tmp)):
    if i%10 == 0:
        final.append(design[k])
        final.append(comments[k])
        k += 1
    final.append(tmp[i])

In [14]:
(final[12:24])

['P2D',
 'P2C',
 'P2Q1O1',
 'P2Q1O2',
 'P2Q2O1',
 'P2Q2O2',
 'P2Q3O1',
 'P2Q3O2',
 'P2Q4O1',
 'P2Q4O2',
 'P2Q5O1',
 'P2Q5O2']

Above are the values collected for each participant. Each participant will have 12 questions and 1 column for each question. Hence, we have 120 columns for 10 participants. Here, the nomenclature is such:
* P1D: For participant1 what is the design preference
* P1C: For participant1 what are the comments
* P1Q1O1: For participant1 question1 option1 what is the rating given
* P1Q1O2: For participant1 question1 option2 what is the rating given

In [15]:
df.columns = final

In [17]:
df.iloc[:5,:20]

P1D   P1C  P1Q1O1  P1Q1O2  P1Q2O1  P1Q2O2  P1Q3O1  P1Q3O2  P1Q4O1  \
1  Option 2  dsds     1.0     1.0     2.0     2.0     2.0     1.0     1.0   
2  Option 1   edc     2.0     2.0     3.0     2.0     2.0     2.0     2.0   
3  Option 2  dsds     1.0     1.0     2.0     2.0     3.0     1.0     1.0   
4  Option 1   edc     2.0     2.0     3.0     2.0     2.0     2.0     2.0   
5  Option 1   edc     1.0     2.0     3.0     3.0     3.0     2.0     2.0   

   P1Q4O2  P1Q5O1  P1Q5O2       P2D      P2C  P2Q1O1  P2Q1O2  P2Q2O1  P2Q2O2  \
1     1.0     3.0     1.0  Option 2    fdfdv     1.0     1.0     3.0     2.0   
2     3.0     1.0     2.0  Option 1  fbb gfg     2.0     2.0     2.0     1.0   
3     1.0     3.0     1.0  Option 2    fdfdv     1.0     1.0     3.0     2.0   
4     3.0     1.0     1.0  Option 1  fbb gfg     2.0     2.0     2.0     1.0   
5     3.0     1.0     2.0  Option 1  fbb gfg     2.0     1.0     2.0     1.0   

   P2Q3O1  P2Q3O2  
1     1.0     2.0  
2     2.0     3.0  
3     1.0     2.0  
4     2.0     3.0  
5     2.0     3.0

In [18]:
optionsdict = {'Option 1':1,'Option 2':2}
for c in df.columns:
    if c[-1] == 'D': df = df.replace({c:optionsdict})
# replace the columns with design choices (ends with 'D') 
# between option 1 & 2 with numeric values (makes understanding easy)

In [19]:
df.iloc[:5,:20]

P1D   P1C  P1Q1O1  P1Q1O2  P1Q2O1  P1Q2O2  P1Q3O1  P1Q3O2  P1Q4O1  P1Q4O2  \
1    2  dsds     1.0     1.0     2.0     2.0     2.0     1.0     1.0     1.0   
2    1   edc     2.0     2.0     3.0     2.0     2.0     2.0     2.0     3.0   
3    2  dsds     1.0     1.0     2.0     2.0     3.0     1.0     1.0     1.0   
4    1   edc     2.0     2.0     3.0     2.0     2.0     2.0     2.0     3.0   
5    1   edc     1.0     2.0     3.0     3.0     3.0     2.0     2.0     3.0   

   P1Q5O1  P1Q5O2  P2D      P2C  P2Q1O1  P2Q1O2  P2Q2O1  P2Q2O2  P2Q3O1  \
1     3.0     1.0    2    fdfdv     1.0     1.0     3.0     2.0     1.0   
2     1.0     2.0    1  fbb gfg     2.0     2.0     2.0     1.0     2.0   
3     3.0     1.0    2    fdfdv     1.0     1.0     3.0     2.0     1.0   
4     1.0     1.0    1  fbb gfg     2.0     2.0     2.0     1.0     2.0   
5     1.0     2.0    1  fbb gfg     2.0     1.0     2.0     1.0     2.0   

   P2Q3O2  
1     2.0  
2     3.0  
3     2.0  
4     3.0  
5     3.0

In [20]:
designmap = pd.read_csv('Participant_Design_Map.csv')
[designmap] = designmap.values.T.tolist()

In [21]:
designmap

[0, 1, 0, 1, 0, 1, 0, 0, 1, 0]

#### Separate the designs with Assistant and designs without Assistant

In [24]:
with_assistant_cols = []
without_assistant_cols = []
k = 0
for c in df.columns:
    # Simply add the cols which donot contain 'O' (the 'D' and 'C' cols)
    if 'O' not in c:
        with_assistant_cols.append(c)
        without_assistant_cols.append(c)
    # condition for design drawn with the assistant
    elif (c.endswith('1') and designmap[k] == 1)\
         or (c.endswith('2') and designmap[k] == 0):
        with_assistant_cols.append(c)
    # condition for design drawn without the assistant
    else:
        without_assistant_cols.append(c)
    
    if (df.columns.get_loc(c) + 1)%(colsPerParticipant) == 0: k += 1

In [25]:
wi_df = df[with_assistant_cols]
wo_df = df[without_assistant_cols]

- *wi_df* is the dataframe for designs drawn with the assistant 
- *wo_df* is the dataframe for designs drawn without the assistant